controllers/edgedeployment_controller_test.go

```diff
@@ -0,0 +1,509 @@
+			ObjectMeta: v1.ObjectMeta{
+				Name:      name,
+				Namespace: "test",
+			},
+			Spec: v1alpha1.EdgeDeviceSpec{
+				OsImageId:   "test",
+				RequestTime: &v1.Time{},
+				Heartbeat:   &v1alpha1.HeartbeatConfiguration{},
+			},
+		}
+	}
+
+	Context("Reconcile", func() {
+		It("Return nil if no edgedeployment found", func() {
+			// given
+			return_erro := errors.NewNotFound(schema.GroupResource{Group: "", Resource: "notfound"}, "notfound")
+			deployRepoMock.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, return_erro).AnyTimes()
+			// when
+			res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+			// then
+			Expect(err).NotTo(HaveOccurred())
+			Expect(res).To(Equal(reconcile.Result{Requeue: false, RequeueAfter: 0}))
+		})
+
+		It("Return error if edgedeployment retrieval failed", func() {
+			// given
+			deployRepoMock.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, fmt.Errorf("Failed")).AnyTimes()
+
+			// when
+			res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+			// then
+			Expect(err).To(HaveOccurred())
+			Expect(res).To(Equal(reconcile.Result{Requeue: true, RequeueAfter: 0}))
+		})
+
+		Context("Finalizers", func() {
+			var (
+				deploymentData *v1alpha1.EdgeDeployment
+			)
+
+			BeforeEach(func() {
+				deploymentData = &v1alpha1.EdgeDeployment{Spec: v1alpha1.EdgeDeploymentSpec{
+					DeviceSelector: &v1.LabelSelector{
+						MatchLabels: map[string]string{"test": "test"},
+					},
+					Device: "test",
+					Type:   "test",
+					Pod:    v1alpha1.Pod{},
+					Data:   &v1alpha1.DataConfiguration{},
+				}}
+
+				deployRepoMock.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).
+					Return(deploymentData, nil).Times(1)
+
+			})
+
+			It("Added finalizer requeue correctly", func() {
+				// given
+				deployRepoMock.EXPECT().Patch(gomock.Any(), deploymentData, gomock.Any()).
+					Return(nil).Do(func(ctx context.Context, old, new *v1alpha1.EdgeDeployment) {
+					Expect(new.Finalizers).To(HaveLen(1))
+					Expect(old.Finalizers).To(HaveLen(0))
+				}).Times(1)
+
+				// when
+				res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+				// then
+				Expect(err).NotTo(HaveOccurred())
+				Expect(res).To(Equal(reconcile.Result{Requeue: true, RequeueAfter: 0}))
+			})
+
+			It("Added finalizer failed requeue with error", func() {
+				// given
+				deployRepoMock.EXPECT().Patch(gomock.Any(), deploymentData, gomock.Any()).
+					Return(nil).Do(func(ctx context.Context, old, new *v1alpha1.EdgeDeployment) {
+					Expect(new.Finalizers).To(HaveLen(1))
+					Expect(old.Finalizers).To(HaveLen(0))
+				}).Return(fmt.Errorf("error")).Times(1)
+
+				// when
+				res, err := edgeDeploymentReconciler.Reconcile(context.TODO(), req)
+
+				// then
+				Expect(err).To(HaveOccurred())
+				Expect(res).To(Equal(reconcile.Result{Requeue: true, RequeueAfter: 0}))
+			})
+		})
+
+		Context("devices section", func() {
```

Comment:
Is this context mainly about using device selector ?
perhaps another context should be for using the device so we'll get full cover of `getMatchingEdgeDevices()`